keps/sig-windows/windows-runtimeclass-support.md

```diff
@@ -0,0 +1,555 @@
+
+With Hyper-V available, Windows can run containers multiple Windows OS versions today, and Linux containers may be available in the future. This document proposes controlling those features through RuntimeClass as well, rather than adding new fields to the Pod API and changing Kubernetes scheduling behavior.
+
+## Motivation
+
+There are a few related customer experience problems when running mixed (multiple-OS and/or multiple-CPU-arch) clusters today. Kubernetes scheduling is based solely on what's declared in the deployment as part of the Pod spec. A Pod could be scheduled to an incompatible node and fail to pull. This error causes a retry and backoff loop, and doesn't automatically fail over to another node.
+
+In addition to the stuck pull failure case, there's also a question of user intent. Some containers are published for multiple OS's and/or architectures - this often referred to as multi-arch but the OCI distribution spec actually includes the OS, OS version, CPU architecture and variant. If a node can run multiple permutations of these, then there's not a deterministic way to know what the user intended to run. Running a container under CPU emulation may have performance penalties, or cost more in terms of leased compute time.
+
+Today with only a single supported Windows version (10.0.17763) & runtime (Docker), the problem is easily mitigated with NodeSelectors or Taints. This is documented today in [Guide for scheduling Windows containers in Kubernetes]. Moving these to RuntimeClasses simplifies the experience further, and can simplify deployment YAML for users in Kubernetes v1.16.
+
+Next, SIG-Windows is adding support for [Windows CRI-ContainerD], Windows nodes will be able to handle running multiple Windows OS versions using Hyper-V isolation. This technology could be used to run Linux containers as well, leading to more ambiguity. This KEP aims to resolve the ambiguity of pulling and running a container image and pod when multiple os/versions/architecture/variants may be supported on a single node.
+
+### Goals
+
+- Schedule a Windows Pod with a compatible host
+  - Allow matching Windows versions without isolation
+  - Allow opt-in on a per-Pod basis to use mismatched Windows versions with compatibility provided by Hyper-V
+- Be able to schedule a specific image from a multi-arch manifest on a given node
+- Provide a simpler experience (fewer lines of YAML) than adding os and version nodeSelector and tolerations to each Pod
+
+### Non-Goals
+
+- Linux container support on Windows is not a requirement or test target for Kubernetes 1.17, but it's not specifically excluded.
+
+## Proposal
+
+For Kubernetes 1.17, we're proposing three key things to improve the experience deploying Windows containers across multiple versions, and enable experimentation with Hyper-V while ContainerD support is in alpha/beta stages.
+
+1. Add `node.kubernetes.io/osversion` label using Windows kubelet
```

Comment:
cc/ @liggitt @mikedanese  on new node label

Reply:
What meaning would this label have on non-windows hosts?

This seems to match the kernel version which was previously rejected, but is actually supported in node-feature-discovery here: https://github.com/kubernetes-sigs/node-feature-discovery#kernel-features

NFD labels are applied centrally though.

Reply:
I see 3 options here:

1. I could set it to kernel version on Linux
2. I could not set it on Linux
3. option 2 plus change the name to `node.kubernetes.io/windowsversion`

Do any of those seem better?

Reply:
The way NFD was written is specific to Linux since it's based off hostvolume paths that don't exist on Windows along with hostNetwork: true. I'd basically have to write an equivalent from scratch to use that strategy.

Reply:
disregard - there's an alternate proposal below of `node.kubernetes.io/windows-build` . Will change to match that suggestion

Reply:
updated to use `windows-build`